README.md
openstack-zvm-compute-anywhere Cookbook
=================================
This cookbook allows z/VM CMA(Cloud Manager Appliance, it provides an easy method to deploy an OpenStack controller or compute node with z/VM support) compute role to be managed by an external OpenStack controller.  
It can only be deployed on z/VM CMA.

Usage
========
Please refer to Chapter 5 in the "Enabling z/VM for OpenStack" user manual from link:
http://www.vm.ibm.com/sysman/openstk.html
 
Testing
=======
Please refer to the [TESTING.md](TESTING.md) for instructions for testing the cookbook.


CHANGELOG.md
Cookbook openstack-zvm-compute-anywhere CHANGELOG
==================================

This file is used to list changes made in each version of the openstack-zvm-compute-anywhere cookbook.

12.0.2
-----
- ZhangShuo - Solve sync issue in compute-cma and ceilometer-cma

12.0.1
-----
- ZhangShuo - Solve sync issue in network-cma

12.0.0
-----
- ZhangShuo - Update to Liberty; Add support to ceilometer; Adopt chefdk to run the chef tests.

10.0.0
-----
- Jichen - Initial release for Juno

- - -
Check the [Markdown Syntax Guide](http://daringfireball.net/projects/markdown/syntax) for help with Markdown.

The [Github Flavored Markdown page](http://github.github.com/github-flavored-markdown/) describes the differences between markdown on github and standard markdown.


spec/spec_helper.rb
# encoding: UTF-8
#
# Cookbook Name:: openstack-zvm-compute-anywhere
# Recipe:: default
#
# Copyright 2015, 2016 IBM Corp.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

require 'chefspec'
require 'chefspec/berkshelf'

require 'chef/application'
require 'securerandom'

ChefSpec::Coverage.start! { add_filter 'openstack-zvm-compute-anywhere' }

::LOG_LEVEL = :fatal
::REDHAT_OPTS = {
  platform: 'redhat',
  version: '6.5',
  log_level: LOG_LEVEL
}

shared_context 'zvm_driver_compute_stubs' do
  before do
    allow_any_instance_of(Chef::Recipe).to receive(:get_user_password)
      .with('xcat', 'xcat-server')
      .and_return('password')
    allow_any_instance_of(Chef::Recipe).to receive(:get_user_password)
      .with('zlinuxroot', 'host')
      .and_return('password')
    allow_any_instance_of(Chef::Recipe).to receive(:get_user_password)
      .with('xcatmnadmin', 'xcat-server')
      .and_return('password')
    allow_any_instance_of(Chef::Recipe).to receive(:get_password)
      .with('db', 'nova')
      .and_return('nova')
    allow_any_instance_of(Chef::Recipe).to receive(:get_password)
      .with('user', 'guest')
      .and_return('mq-guess')
    allow_any_instance_of(Chef::Recipe).to receive(:get_password)
      .with('user', 'admin')
      .and_return('admin')
    allow_any_instance_of(Chef::Recipe).to receive(:get_password)
      .with('service', 'openstack-network')
      .and_return('neutron')
    allow_any_instance_of(Chef::Recipe).to receive(:get_password)
      .with('service', 'openstack-compute')
      .and_return('nova')
    allow_any_instance_of(Chef::Recipe).to receive(:get_password)
      .with('service', 'openstack-bare-metal')
      .and_return('ironic-pass')
    allow_any_instance_of(Chef::Recipe).to receive(:get_password)
      .with('token', 'neutron_metadata_secret')
      .and_return('metadata-secret')
    allow_any_instance_of(Chef::Recipe).to receive(:get_password)
      .with('token', 'openstack_vmware_secret_name')
      .and_return 'vmware_secret_name'
    allow(::Net::SSH).to receive(:start)
  end
end

shared_context 'zvm_driver_network_stubs' do
  before do
    allow_any_instance_of(Chef::Recipe).to receive(:get_password)
      .with('db', 'neutron')
      .and_return('neutron')
    allow_any_instance_of(Chef::Recipe).to receive(:get_password)
      .with('service', 'openstack-network')
      .and_return('openstack-network')
    allow_any_instance_of(Chef::Recipe).to receive(:get_user_password)
      .with('xcat', 'xcat-server')
      .and_return('password')
    allow_any_instance_of(Chef::Recipe).to receive(:get_password)
      .with('user', 'guest')
      .and_return('guest')
  end
end

shared_context 'zvm_driver_ceilometer_stubs' do
  before do
    allow_any_instance_of(Chef::Recipe).to receive(:get_password)
      .with('db', 'ceilometer')
      .and_return('ceilometer')
    allow_any_instance_of(Chef::Recipe).to receive(:get_password)
      .with('user', 'guest')
      .and_return('rabbitmq')
    allow_any_instance_of(Chef::Recipe).to receive(:get_password)
      .with('user', 'qpid')
      .and_return('qpid')
    allow_any_instance_of(Chef::Recipe).to receive(:get_password)
      .with('service', 'openstack-ceilometer')
      .and_return('ceilometer')
    allow_any_instance_of(Chef::Recipe).to receive(:get_password)
      .with('token', 'openstack_metering_secret')
      .and_return('ceilometer')
    allow_any_instance_of(Chef::Recipe).to receive(:get_password)
      .with('token', 'openstack_vmware_secret_name')
      .and_return('vmware_secret_name')
  end
end

shared_context 'compute_common_stubs' do
  before do
    allow_any_instance_of(Chef::Recipe).to receive(:rabbit_servers)
      .and_return '1.1.1.1:5672,2.2.2.2:5672'
    allow_any_instance_of(Chef::Recipe).to receive(:address_for)
      .with('lo')
      .and_return '127.0.1.1'
    allow_any_instance_of(Chef::Recipe).to receive(:search_for)
      .with('os-identity').and_return(
        [{
          'openstack' => {
            'identity' => {
              'admin_tenant_name' => 'admin',
              'admin_user' => 'admin'
            }
          }
        }]
      )
    allow_any_instance_of(Chef::Recipe).to receive(:get_secret)
      .with('openstack_identity_bootstrap_token')
      .and_return('bootstrap-token')
    allow_any_instance_of(Chef::Recipe).to receive(:get_secret)
      .with('neutron_metadata_secret')
      .and_return('metadata-secret')
    allow_any_instance_of(Chef::Recipe).to receive(:get_secret) # this is the rbd_uuid default name
      .with('rbd_secret_uuid')
      .and_return '00000000-0000-0000-0000-000000000000'
    allow_any_instance_of(Chef::Recipe).to receive(:get_secret)
      .with('openstack_vmware_secret_name')
      .and_return 'vmware_secret_name'
    allow_any_instance_of(Chef::Recipe).to receive(:get_password)
      .with('db', anything)
      .and_return('')
    allow_any_instance_of(Chef::Recipe).to receive(:get_password)
      .with('user', 'guest')
      .and_return('mq-pass')
    allow_any_instance_of(Chef::Recipe).to receive(:get_password)
      .with('user', 'admin')
      .and_return('admin')
    allow_any_instance_of(Chef::Recipe).to receive(:get_password)
      .with('service', 'openstack-compute')
      .and_return('nova-pass')
    allow_any_instance_of(Chef::Recipe).to receive(:get_password)
      .with('service', 'openstack-network')
      .and_return('neutron-pass')
    allow_any_instance_of(Chef::Recipe).to receive(:get_password)
      .with('service', 'rbd_block_storage')
      .and_return 'cinder-rbd-pass'
    allow_any_instance_of(Chef::Recipe).to receive(:memcached_servers).and_return []
    allow_any_instance_of(Chef::Recipe).to receive(:system)
      .with("grub2-set-default 'openSUSE GNU/Linux, with Xen hypervisor'")
      .and_return(true)
    allow(Chef::Application).to receive(:fatal!)
    allow(SecureRandom).to receive(:hex) { 'ad3313264ea51d8c6a3d1c5b140b9883' }
    stub_command('nova-manage network list | grep 192.168.100.0/24').and_return(false)
    stub_command('nova-manage network list | grep 192.168.200.0/24').and_return(false)
    stub_command("nova-manage floating list |grep -E '.*([0-9]{1,3}[.]){3}[0-9]{1,3}*'").and_return(false)
    stub_command('virsh net-list | grep -q default').and_return(true)
    stub_command('ovs-vsctl br-exists br-int').and_return(true)
    stub_command('ovs-vsctl br-exists br-tun').and_return(true)
    stub_command('virsh secret-list | grep 00000000-0000-0000-0000-000000000000').and_return(false)
    stub_command("virsh secret-get-value 00000000-0000-0000-0000-000000000000 | grep 'cinder-rbd-pass'").and_return(false)
  end
end

shared_context 'network-common-stubs' do
  before do
    allow_any_instance_of(Chef::Recipe).to receive(:rabbit_servers)
      .and_return('1.1.1.1:5672,2.2.2.2:5672')
    allow_any_instance_of(Chef::Recipe).to receive(:config_by_role)
      .with('rabbitmq-server', 'queue').and_return(
        host: 'rabbit-host',
        port: 'rabbit-port'
      )
    allow_any_instance_of(Chef::Recipe).to receive(:config_by_role)
      .with('glance-api', 'glance').and_return []
    allow_any_instance_of(Chef::Recipe).to receive(:secret)
      .with('secrets', 'openstack_identity_bootstrap_token')
      .and_return('bootstrap-token')
    allow_any_instance_of(Chef::Recipe).to receive(:secret)
      .with('secrets', 'neutron_metadata_secret')
      .and_return('metadata-secret')
    allow_any_instance_of(Chef::Recipe).to receive(:get_password)
      .with('db', anything)
      .and_return('neutron')
    allow_any_instance_of(Chef::Recipe).to receive(:get_password)
      .with('service', 'openstack-network')
      .and_return('neutron-pass')
    allow_any_instance_of(Chef::Recipe).to receive(:get_password)
      .with('user', 'guest')
      .and_return('mq-pass')
    allow(Chef::Application).to receive(:fatal!)
    allow_any_instance_of(Chef::Recipe).to receive(:get_password)
      .with('service', 'openstack-compute')
      .and_return('nova-pass')
    allow_any_instance_of(Chef::Resource::RubyBlock).to receive(:openstack_command_env)
      .with('admin', 'admin')
      .and_return({})
    allow_any_instance_of(Chef::Resource::RubyBlock).to receive(:identity_uuid)
      .with('tenant', 'name', 'service', {})
      .and_return('000-UUID-FROM-CLI')

    stub_command('dpkg -l | grep openvswitch-switch | grep 1.10.2-1').and_return(true)
    stub_command('ovs-vsctl br-exists br-int').and_return(false)
    stub_command('ovs-vsctl br-exists br-tun').and_return(false)
    stub_command('ip link show eth1').and_return(false)
  end
end


spec/compute_spec.rb
# encoding: UTF-8
#
# Cookbook Name:: openstack-zvm-compute-anywhere
# Recipe:: compute-cma
#
# Copyright 2015, 2016 IBM Corp.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

require_relative 'spec_helper'

describe 'openstack-zvm-compute-anywhere::compute-cma' do
  describe 'redhat' do
    include_context 'compute_common_stubs'
    include_context 'zvm_driver_compute_stubs'
    let(:runner) { ChefSpec::ServerRunner.new(REDHAT_OPTS) }
    let(:node) { runner.node }
    let(:chef_run) do
      shellout = double
      cmd = double
      cho = double
      allow(Mixlib::ShellOut).to receive(:new).and_return(shellout)
      allow(shellout).to receive(:run_command).and_return(cmd)
      allow(cmd).to receive(:stdout).and_return(cho)
      allow(cho).to receive(:chomp).and_return('~nova')
      node.set['ibm-openstack']['zvm-appliance']['unchangeable-confs']['nova-conf']['default']['host'] = 'host'
      node.set['ibm-openstack']['zvm-appliance']['unchangeable-confs']['nova-conf']['default']['zvm_diskpool_type'] = 'ECKD'
      node.set['ibm-openstack']['zvm-driver']['config']['ram_allocation_ratio'] = 10
      node.set['ibm-openstack']['zvm-appliance']['unchangeable-confs']['mnadmin_user'] = 'mnadmin'
      runner.converge(described_recipe)
    end

    it 'check nova conf' do
      expect(chef_run).to create_template('/etc/nova/nova.conf').with(
        owner: 'nova',
        group: 'nova',
        mode: 00640
      )
      expect(chef_run).to render_file('/etc/nova/nova.conf').with_content('compute_driver=nova.virt.zvm.ZVMDriver')
      expect(chef_run).to render_file('/etc/nova/nova.conf').with_content('ram_allocation_ratio=10')
      expect(chef_run).to render_file('/etc/nova/nova.conf').with_content('zvm_diskpool_type=ECKD')
      expect(chef_run).to render_file('/etc/nova/nova.conf').with_content('config_drive_format=iso9660')
      expect(chef_run).to render_file('/etc/nova/nova.conf').with_content('host=host')
      expect(chef_run).to render_file('/etc/nova/nova.conf').with_content('admin_password = nova')
    end

    it 'includes openstack-common::openrc recipe' do
      expect(chef_run).to include_recipe('openstack-common::openrc')
    end

    it 'enables or starts nova-compute on boot' do
      expect(chef_run).to enable_service('openstack-nova-compute')
      expect(chef_run).to start_service('openstack-nova-compute')
    end

    it 'openrc force_override attributes' do
      expect(chef_run.node['openstack']['openrc']['path']).to eq('/home/mnadmin')
      expect(chef_run.node['openstack']['openrc']['file']).to eq('openrc')
      expect(chef_run.node['openstack']['openrc']['user']).to eq('mnadmin')
      expect(chef_run.node['openstack']['openrc']['group']).to eq('root')
      expect(chef_run.node['openstack']['openrc']['file_mode']).to eq('0600')
      expect(chef_run.node['openstack']['openrc']['path_mode']).to eq('0755')
    end

  end
end


spec/network_spec.rb
# encoding: UTF-8
#
# Cookbook Name:: openstack-zvm-compute-anywhere
# Recipe:: network-cma
#
# Copyright 2015, 2016 IBM Corp.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

require_relative 'spec_helper'

describe 'openstack-zvm-compute-anywhere::network-cma' do
  include_context 'network-common-stubs'
  describe 'redhat' do
    let(:runner) { ChefSpec::ServerRunner.new(REDHAT_OPTS) }
    let(:node) { runner.node }
    let(:chef_run) do
      node.set['openstack']['compute']['network']['service_type'] = 'neutron'
      node.set['openstack']['network']['rpc_thread_pool_size'] = 1000
      runner.converge(described_recipe)
    end

    it 'check neutron conf' do
      expect(chef_run).to create_template('/etc/neutron/neutron.conf').with(
        owner: 'neutron',
        group: 'neutron',
        mode: 00640
      )

      expect(chef_run).to render_file('/etc/neutron/neutron.conf').with_content('rpc_thread_pool_size = 1000')
      expect(chef_run).to render_file('/etc/neutron/neutron.conf').with_content('admin_password = neutron-pass')
    end

    it 'starts neutron-zvm-agent' do
      expect(chef_run).to start_service('neutron-zvm-agent')
    end
  end
end


metadata.rb
name             'openstack-zvm-compute-anywhere'
maintainer       'IBM Corp.'
maintainer_email 'https://github.com/zVM-Cloud'
license          'Apache 2.0'
description      'Installs/Configures openstack-zvm-compute-anywhere'
long_description IO.read(File.join(File.dirname(__FILE__), 'README.md'))
version          '12.0.2'

recipe 'openstack-zvm-compute-anywhere::compute-cma', 'Reconfigures nova on CMA to make it work with external controller'
recipe 'openstack-zvm-compute-anywhere::network-cma', 'Reconfigures neutron on CMA to make it work with external controller'
recipe 'openstack-zvm-compute-anywhere::ceilometer-cma', 'Reconfigures ceilometer on CMA to make it work with external controller'

supports 'redhat'

depends 'openstack-common', '>= 12.0.0'
depends 'openstack-compute', '>= 12.0.0'
depends 'openstack-network', '>= 12.0.0'
depends 'openstack-telemetry', '>= 12.0.0'


recipes/network-cma.rb
# encoding: UTF-8
#
# Cookbook Name:: openstack-zvm-compute-anywhere
# Recipe:: network-cma
#
# Copyright 2015, 2016 IBM Corp.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

db_type = node['openstack']['db']['network']['service_type']
node.default['openstack']['db']['python_packages'].delete(db_type)
node.force_override['openstack']['db']['python_packages'][db_type] = [] # :pragma-foodcritic: ~FC019
%w{nova_network_packages neutron_packages neutron_lb_packages neutron_vpn_packages neutron_client_packages}.each do |attr|
  node.default['openstack']['network']['platform'].delete(attr)
  node.force_override['openstack']['network']['platform'][attr] = [] # :pragma-foodcritic: ~FC019
end
node.default['openstack']['identity']['platform'].delete('keystone_client_packages')
node.force_override['openstack']['identity']['platform']['keystone_client_packages'] = [] # :pragma-foodcritic: ~FC019

include_recipe 'openstack-network::default'

run_context.resource_collection.each do |r|
  r.action(:nothing) if r.to_s == 'directory[/etc/neutron/plugins]'
  r.action(:nothing) if r.to_s == 'directory[/var/cache/neutron]'
  r.action(:nothing) if r.to_s == 'template[/etc/neutron/rootwrap.conf]'
  r.action(:nothing) if r.to_s == 'service[neutron-server]'
  r.action(:nothing) if r.to_s == 'directory[/etc/neutron/plugins/ml2]'
  r.action(:nothing) if r.to_s == 'template[/etc/neutron/plugins/ml2/ml2_conf.ini]'
  r.action(:nothing) if r.to_s == 'link[/etc/neutron/plugin.ini]'
  r.action(:nothing) if r.to_s == 'template[/etc/default/neutron-server]'
end

service 'neutron-zvm-agent' do
  supports status: true, restart: true
  action [:start]
  subscribes :restart, 'template[/etc/neutron/neutron.conf]', :delayed
end


recipes/ceilometer-cma.rb
# encoding: UTF-8
#
# Cookbook Name:: openstack-zvm-compute-anywhere
# Recipe:: ceilometer-cma
#
# Copyright 2015, 2016 IBM Corp.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

db_type = node['openstack']['db']['telemetry']['service_type']
node.default['openstack']['db']['python_packages'].delete(db_type)
node.force_override['openstack']['db']['python_packages'][db_type] = [] # :pragma-foodcritic: ~FC019
node.default['openstack']['telemetry']['platform'].delete('common_packages')
node.force_override['openstack']['telemetry']['platform']['common_packages'] = [] # :pragma-foodcritic: ~FC019
node.default['openstack']['telemetry'].delete('hypervisor_inspector')
node.force_override['openstack']['telemetry']['hypervisor_inspector'] = 'zvm' # :pragma-foodcritic: ~FC019

telemetry_conf_dir = node['openstack']['telemetry']['conf_dir']
telemetry_conf_template = node['openstack']['telemetry']['conf']

include_recipe 'openstack-telemetry::common'

run_context.resource_collection.each do |r|
  r.action(:nothing) if r.to_s == "directory[#{telemetry_conf_dir}]"
  r.action(:nothing) if r.to_s == "template[#{telemetry_conf_template}]"
end

parent_variables = run_context.resource_collection.find("template[#{telemetry_conf_template}]").variables

template telemetry_conf_template do
  source 'ceilometer_zvm.conf.erb'
  owner node['openstack']['telemetry']['user']
  group node['openstack']['telemetry']['group']
  mode 00640

  variables(
    parent_variables
  )
end

service 'ceilometer-agent-compute' do
  service_name 'openstack-ceilometer-polling'
  supports status: true, restart: true
  subscribes :restart, "template[#{node['openstack']['telemetry']['conf']}]"

  action [:start]
end


recipes/compute-cma.rb
# encoding: UTF-8
#
# Cookbook Name:: openstack-zvm-compute-anywhere
# Recipe:: compute-cma
#
# Copyright 2015, 2016 IBM Corp.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

# generate openrc file
mnadmin = node['ibm-openstack']['zvm-appliance']['unchangeable-confs']['mnadmin_user']
node.force_override['openstack']['openrc']['path'] = "/home/#{mnadmin}" # :pragma-foodcritic: ~FC019
node.force_override['openstack']['openrc']['file'] = 'openrc' # :pragma-foodcritic: ~FC019
node.force_override['openstack']['openrc']['user'] = mnadmin # :pragma-foodcritic: ~FC019
node.force_override['openstack']['openrc']['group'] = 'root' # :pragma-foodcritic: ~FC019
node.force_override['openstack']['openrc']['file_mode'] = '0600' # :pragma-foodcritic: ~FC019
node.force_override['openstack']['openrc']['path_mode'] = '0755' # :pragma-foodcritic: ~FC019
include_recipe 'openstack-common::openrc'

node.default['openstack']['compute']['platform'].delete('common_packages')
node.force_override['openstack']['compute']['platform']['common_packages'] = [] # :pragma-foodcritic: ~FC019
db_type = node['openstack']['db']['compute']['service_type']
node.default['openstack']['db']['python_packages'].delete(db_type)
node.force_override['openstack']['db']['python_packages'][db_type] = [] # :pragma-foodcritic: ~FC019
node.default['openstack']['compute']['platform'].delete('memcache_python_packages')
node.force_override['openstack']['compute']['platform']['memcache_python_packages'] = [] # :pragma-foodcritic: ~FC019

state_path = node['openstack']['compute']['state_path']
lock_path = node['openstack']['compute']['lock_path']

include_recipe 'openstack-compute::nova-common'

run_context.resource_collection.each do |r|
  r.action(:nothing) if r.to_s == "directory[/etc/nova]"
  r.action(:nothing) if r.to_s == "directory[#{state_path}]"
  r.action(:nothing) if r.to_s == "directory[#{lock_path}]"
  r.action(:nothing) if r.to_s == "template[/etc/nova/nova.conf]}]"
  r.action(:nothing) if r.to_s == "template[/etc/nova/rootwrap.conf]}]"
  r.action(:nothing) if r.to_s == "execute[enable nova login]"
end

parent_variables = run_context.resource_collection.find("template[/etc/nova/nova.conf]").variables

node.default['openstack']['compute'].delete('driver')
node.default['openstack']['compute']['config'].delete('force_config_drive')
node.default['openstack']['compute']['config'].delete('config_drive_format')
node.default['openstack']['compute']['config'].delete('ram_allocation_ratio')
node.default['openstack']['compute'].delete('rpc_response_timeout')
node.force_override['openstack']['compute']['driver'] = 'nova.virt.zvm.ZVMDriver' # :pragma-foodcritic: ~FC019
node.force_override['openstack']['compute']['config']['force_config_drive'] = true # :pragma-foodcritic: ~FC019
node.force_override['openstack']['compute']['config']['config_drive_format'] = 'iso9660' # :pragma-foodcritic: ~FC019
node.force_override['openstack']['compute']['config']['ram_allocation_ratio'] = node['ibm-openstack']['zvm-driver']['config']['ram_allocation_ratio'] # :pragma-foodcritic: ~FC019
node.force_override['openstack']['compute']['rpc_response_timeout'] = node['ibm-openstack']['zvm-driver']['rpc_response_timeout'] # :pragma-foodcritic: ~FC019

template '/etc/nova/nova.conf' do
  source 'nova_zvm.conf.erb'
  owner node['openstack']['compute']['user']
  group node['openstack']['compute']['group']
  mode 00640

  variables(
    parent_variables
  )
end

compute_service = node['openstack']['compute']['platform']['compute_compute_service']

service compute_service do
  supports status: true, restart: true
  provider Chef::Provider::Service::Init::Redhat
  subscribes :restart, resources('template[/etc/nova/nova.conf]')

  action [:start]
end
